<left>
    <a href=img><img src = "https://proetica.icts.com.br/assets/img/ICTS-Protiviti-Logo.png" width="400"  />
</left>

## Processo de extração, limpeza e preparação de dataset referente aos dados de processos de
## tratatamento térmico (recozimento)

## 
Autor: Tiago Moraes

### 1. Carregamento de informações

In [1]:
import pandas as pd
import numpy as np
print ("Libraries sucessfully loaded") # print para assegurar que as bibliotecas necessárias carregaram corretamente

Libraries sucessfully loaded


In [2]:
# carregando arquivo data_experimentos, hospedado em repostório no Github para facilitar a compatibilidade deste notebook ao ser compartilhado 
filename_exp = "https://raw.githubusercontent.com/tiagottmoraes/admission-test/main/datasets/data_experimentos.csv"

In [3]:
# criado dataframe para armazenar os dados obtivos do arquivo .csv
df = pd.read_csv(filename_exp)
df

exp_id   ano  mes
0          2  2020    8
1       1368  2020    8
2       1130  2020    8
3        402  2020    8
4        407  2020    8
...      ...   ...  ...
2419    1354  2020    7
2420    1610  2020    7
2421    2243  2020    7
2422     900  2020    7
2423     534  2020    7

[2424 rows x 3 columns]

In [4]:
# carregando arquivo train.csv 
filename_train = "https://raw.githubusercontent.com/tiagottmoraes/admission-test/main/datasets/train.csv"

In [5]:
# criado um dataframe para armazenar os dados do arquivo carregado
other = pd.read_csv(filename_train)

### 2. Juntando os dois datasets



In [6]:
# Antes de proceder com a junção dos dataframes, checou-se se o tamanho de ambas as bases para entender como o processo deveria ser feito
other.shape

(2424, 40)

In [7]:
# junção
dados = df.join(other)
dados

exp_id   ano  mes family product-type steel  carbon  hardness  \
0          2  2020    8      ?            C     A       8         0   
1       1368  2020    8      ?            C     R       0         0   
2       1130  2020    8      ?            C     A       0        60   
3        402  2020    8      ?            C     A       0        60   
4        407  2020    8      ?            C     A       0        45   
...      ...   ...  ...    ...          ...   ...     ...       ...   
2419    1354  2020    7      ?            C     R       0         0   
2420    1610  2020    7     TN            C     A       0         0   
2421    2243  2020    7      ?            C     R       0         0   
2422     900  2020    7      ?            C     A       0        85   
2423     534  2020    7      ?            C     A       0        85   

     temper_rolling condition  ...  p  shape      thick       width       len  \
0                 ?         S  ...  ?   COIL     _0.7_     _610.0_       _0_    
1                 ?         S  ...  ?  SHEET     _0.7_    _1300.0_     _762_    
2                 T         ?  ...  ?   COIL   _2.801_     _385.1_       _0_    
3                 T         ?  ...  ?  SHEET   _0.801_     _255.0_     _269_    
4                 ?         S  ...  ?   COIL     _1.6_     _610.0_       _0_    
...             ...       ...  ... ..    ...        ...         ...       ...   
2419              ?         S  ...  ?  SHEET     _1.6_    _1500.0_    _4170_    
2420              ?         ?  ...  ?  SHEET     _1.2_     _609.9_       _1_    
2421              ?         S  ...  ?  SHEET   _1.601_     _830.0_     _880_    
2422              T         ?  ...  ?   COIL     _0.4_      _20.0_       _0_    
2423              T         ?  ...  ?   COIL     _4.0_     _610.0_       _0_    

     oil bore packing recozimento experimento  
0      ?    0       ?        ruim       exp_2  
1      ?    0       ?        ruim    exp_1368  
2      ?    0       ?       ideal    exp_1130  
3      ?    0       ?       ideal     exp_402  
4      ?    0       ?     mediano     exp_407  
...   ..  ...     ...         ...         ...  
2419   ?    0       ?     mediano    exp_1354  
2420   ?    0       ?     mediano    exp_1610  
2421   ?    0       ?     mediano    exp_2243  
2422   ?    0       ?     mediano     exp_900  
2423   ?  500       ?     mediano     exp_534  

[2424 rows x 43 columns]

Observando o dataframe resultante da junção, observa-se que a primeira coluna ('exp_id') e a última ('experimento') apresentam a mesma informação, ou seja, o id de cada experimento realizado. Assim, optou-se por remover a última coluna deste df.

In [8]:
del dados['experimento']

In [9]:
dados.head()

exp_id   ano  mes family product-type steel  carbon  hardness  \
0       2  2020    8      ?            C     A       8         0   
1    1368  2020    8      ?            C     R       0         0   
2    1130  2020    8      ?            C     A       0        60   
3     402  2020    8      ?            C     A       0        60   
4     407  2020    8      ?            C     A       0        45   

  temper_rolling condition  ...  s  p  shape      thick       width      len  \
0              ?         S  ...  ?  ?   COIL     _0.7_     _610.0_      _0_    
1              ?         S  ...  ?  ?  SHEET     _0.7_    _1300.0_    _762_    
2              T         ?  ...  ?  ?   COIL   _2.801_     _385.1_      _0_    
3              T         ?  ...  ?  ?  SHEET   _0.801_     _255.0_    _269_    
4              ?         S  ...  ?  ?   COIL     _1.6_     _610.0_      _0_    

  oil bore packing recozimento  
0   ?    0       ?        ruim  
1   ?    0       ?        ruim  
2   ?    0       ?       ideal  
3   ?    0       ?       ideal  
4   ?    0       ?     mediano  

[5 rows x 42 columns]

### 3. Formatação e limpeza dos dados

In [10]:
# as colunas "thick", "width" e "len" estão formatadas incorretamente, com o caracter '_' antes e depois dos valores. Utilizou-se a função "replace" para remover este caracter
dados["thick"] = dados["thick"].replace(to_replace ="_", value = '', regex = True)
dados["width"] = dados["width"].replace(to_replace ="_", value = '', regex = True)
dados["len"] = dados["len"].replace(to_replace ="_", value = '', regex = True)
dados

exp_id   ano  mes family product-type steel  carbon  hardness  \
0          2  2020    8      ?            C     A       8         0   
1       1368  2020    8      ?            C     R       0         0   
2       1130  2020    8      ?            C     A       0        60   
3        402  2020    8      ?            C     A       0        60   
4        407  2020    8      ?            C     A       0        45   
...      ...   ...  ...    ...          ...   ...     ...       ...   
2419    1354  2020    7      ?            C     R       0         0   
2420    1610  2020    7     TN            C     A       0         0   
2421    2243  2020    7      ?            C     R       0         0   
2422     900  2020    7      ?            C     A       0        85   
2423     534  2020    7      ?            C     A       0        85   

     temper_rolling condition  ...  s  p  shape    thick     width     len  \
0                 ?         S  ...  ?  ?   COIL     0.7     610.0       0    
1                 ?         S  ...  ?  ?  SHEET     0.7    1300.0     762    
2                 T         ?  ...  ?  ?   COIL   2.801     385.1       0    
3                 T         ?  ...  ?  ?  SHEET   0.801     255.0     269    
4                 ?         S  ...  ?  ?   COIL     1.6     610.0       0    
...             ...       ...  ... .. ..    ...      ...       ...     ...   
2419              ?         S  ...  ?  ?  SHEET     1.6    1500.0    4170    
2420              ?         ?  ...  ?  ?  SHEET     1.2     609.9       1    
2421              ?         S  ...  ?  ?  SHEET   1.601     830.0     880    
2422              T         ?  ...  ?  ?   COIL     0.4      20.0       0    
2423              T         ?  ...  ?  ?   COIL     4.0     610.0       0    

     oil bore packing recozimento  
0      ?    0       ?        ruim  
1      ?    0       ?        ruim  
2      ?    0       ?       ideal  
3      ?    0       ?       ideal  
4      ?    0       ?     mediano  
...   ..  ...     ...         ...  
2419   ?    0       ?     mediano  
2420   ?    0       ?     mediano  
2421   ?    0       ?     mediano  
2422   ?    0       ?     mediano  
2423   ?  500       ?     mediano  

[2424 rows x 42 columns]

#### 3.1 Análise de valores ausentes
Anteriormente observou-se a presença de diversos caracteres incorretos (como "?"). Assim, procedeu-se com a análise de valores ausentes ou incorretos

In [11]:
# trocando todos os "?" por NaN com função do numpy
dados.replace("?",np.nan, inplace=True)
dados.head()

exp_id   ano  mes family product-type steel  carbon  hardness  \
0       2  2020    8    NaN            C     A       8         0   
1    1368  2020    8    NaN            C     R       0         0   
2    1130  2020    8    NaN            C     A       0        60   
3     402  2020    8    NaN            C     A       0        60   
4     407  2020    8    NaN            C     A       0        45   

  temper_rolling condition  ...   s   p  shape    thick     width    len  oil  \
0            NaN         S  ... NaN NaN   COIL     0.7     610.0      0   NaN   
1            NaN         S  ... NaN NaN  SHEET     0.7    1300.0    762   NaN   
2              T       NaN  ... NaN NaN   COIL   2.801     385.1      0   NaN   
3              T       NaN  ... NaN NaN  SHEET   0.801     255.0    269   NaN   
4            NaN         S  ... NaN NaN   COIL     1.6     610.0      0   NaN   

  bore packing recozimento  
0    0     NaN        ruim  
1    0     NaN        ruim  
2    0     NaN       ideal  
3    0     NaN       ideal  
4    0     NaN     mediano  

[5 rows x 42 columns]

In [12]:
# verificando se existem células sem valor
missing_data=dados.isnull()
missing_data.head()

exp_id    ano    mes  family  product-type  steel  carbon  hardness  \
0   False  False  False    True         False  False   False     False   
1   False  False  False    True         False  False   False     False   
2   False  False  False    True         False  False   False     False   
3   False  False  False    True         False  False   False     False   
4   False  False  False    True         False  False   False     False   

   temper_rolling  condition  ...     s     p  shape  thick  width    len  \
0            True      False  ...  True  True  False  False  False  False   
1            True      False  ...  True  True  False  False  False  False   
2           False       True  ...  True  True  False  False  False  False   
3           False       True  ...  True  True  False  False  False  False   
4            True      False  ...  True  True  False  False  False  False   

    oil   bore  packing  recozimento  
0  True  False     True        False  
1  True  False     True        False  
2  True  False     True        False  
3  True  False     True        False  
4  True  False     True        False  

[5 rows x 42 columns]

In [13]:
print(" \nTotal de itens NaN presentes nas colunas do DataFrame 'Dados': \n\n", dados.isnull().sum())


 
Total de itens NaN presentes nas colunas do DataFrame 'Dados': 

 exp_id                       0
ano                          0
mes                          0
family                    2073
product-type                 0
steel                      231
carbon                       0
hardness                     0
temper_rolling            2043
condition                  822
formability                873
strength                     0
non-ageing                2139
surface-finish            2400
surface-quality            645
enamelability             2385
bc                        2421
bf                        2070
bt                        2220
bw/me                     1851
bl                        2019
m                         2424
chrom                     2352
phos                      2406
cbond                     2223
marvi                     2424
exptl                     2418
ferro                     2343
corr                      2424
blue/bright/varn/clean    2409
lu

Observa-se que uma grande quantidade de variáveis possuem a totalidade de suas linhas (ou seja 2424) como NaN. Além disso, diversas outras variáveis possuem mais de 90% de seu conteúdo como sendo NaN. Em ambos os casos, optou-se por remover estas variáveis, uma vez que elas possuem pouco valor estatístico, considerando o espaço amostral, além de que a sua remoção deixaria o dataframe muito mais enxuto. 

In [14]:
df_clean = dados.drop(['family','temper_rolling','non-ageing','surface-finish','enamelability', 'bc', 'bw/me', 'bf', 'bt', 'bl', 'chrom', 'phos', 'cbond', 'exptl', 'ferro', 'blue/bright/varn/clean','lustre', 'oil', 'packing','m', 'marvi', 'corr' , 'jurofm' , 's' , 'p'], axis = 1)

#### 3.2 Ajustando o tipo de cada variável do DF

In [15]:
# obsevando os tipos de variáveis de cada coluna 
df_clean.dtypes

exp_id              int64
ano                 int64
mes                 int64
product-type       object
steel              object
carbon              int64
hardness            int64
condition          object
formability        object
strength            int64
surface-quality    object
shape              object
thick              object
width              object
len                object
bore                int64
recozimento        object
dtype: object

Observa-se que diversas variáveis estão indereçadas incorretamente, como por exemplo, "thick" e "width" que estão como uma string (object), mas deveriam ser um float

In [16]:
df_clean[["formability", "thick", "width", "len"]] = df_clean[["formability", "thick", "width", "len"]].astype("float")
df_clean[['mes', 'ano', 'exp_id']] = df_clean[['mes','ano','exp_id']].astype('float')
df_clean.dtypes

exp_id             float64
ano                float64
mes                float64
product-type        object
steel               object
carbon               int64
hardness             int64
condition           object
formability        float64
strength             int64
surface-quality     object
shape               object
thick              float64
width              float64
len                float64
bore                 int64
recozimento         object
dtype: object

#### Como solicitado, separou-se os valores referentes ao mes de agosto (mes 8) para a análise completa do histórico dos dados

In [17]:
grouped = df_clean.groupby(df_clean.mes)
df_clean = grouped.get_group(8)
df_clean

exp_id     ano  mes product-type steel  carbon  hardness condition  \
0        2.0  2020.0  8.0            C     A       8         0         S   
1     1368.0  2020.0  8.0            C     R       0         0         S   
2     1130.0  2020.0  8.0            C     A       0        60       NaN   
3      402.0  2020.0  8.0            C     A       0        60       NaN   
4      407.0  2020.0  8.0            C     A       0        45         S   
...      ...     ...  ...          ...   ...     ...       ...       ...   
1611   145.0  2019.0  8.0            C     R       0         0         S   
1612   937.0  2019.0  8.0            C     A       0         0       NaN   
1613  2228.0  2019.0  8.0            C     R       0         0         S   
1614  1263.0  2019.0  8.0            C     A       0        85       NaN   
1615  1236.0  2019.0  8.0            C     A       0        85       NaN   

      formability  strength surface-quality  shape  thick   width     len  \
0             NaN         0               G   COIL  0.700   610.0     0.0   
1             2.0         0               E  SHEET  0.700  1300.0   762.0   
2             NaN         0               G   COIL  2.801   385.1     0.0   
3             NaN         0               G  SHEET  0.801   255.0   269.0   
4             NaN         0               D   COIL  1.600   610.0     0.0   
...           ...       ...             ...    ...    ...     ...     ...   
1611          2.0         0             NaN  SHEET  1.600  1500.0  4170.0   
1612          3.0         0             NaN  SHEET  1.200   609.9     1.0   
1613          3.0         0             NaN  SHEET  1.601   830.0   880.0   
1614          NaN         0               G   COIL  0.400    20.0     0.0   
1615          NaN         0               G   COIL  4.000   610.0     0.0   

      bore recozimento  
0        0        ruim  
1        0        ruim  
2        0       ideal  
3        0       ideal  
4        0     mediano  
...    ...         ...  
1611     0       ideal  
1612     0       ideal  
1613     0       ideal  
1614     0       ideal  
1615   500       ideal  

[1616 rows x 17 columns]

Observe que agora o dataframe possui 1616 linhas

In [18]:
print(" \nTotal de itens NaN presentes nas colunas do DataFrame 'Dados' : \n\n", df_clean.isnull().sum())

 
Total de itens NaN presentes nas colunas do DataFrame 'Dados' : 

 exp_id               0
ano                  0
mes                  0
product-type         0
steel              154
carbon               0
hardness             0
condition          548
formability        582
strength             0
surface-quality    430
shape                0
thick                0
width                0
len                  0
bore                 0
recozimento          0
dtype: int64


Algumas variáveis ainda possuem células com NaN

#### 3.3 Preenchendo espaços vazios com valores médios ou mais frequentes
Para as variáveis com uma menor quantidade de valores vazios, optou-se por preenche-los com a média (no caso de variáveis contínuas) ou pela moda (no caso de variáveis discretas)

In [19]:
# encontrado os elementos mais frequentes
print ("Most frequent element present in 'steel' column:", df_clean['steel'].value_counts().idxmax() )
print ("Most frequent element present in 'condition' column:", df_clean['condition'].value_counts().idxmax() )
print ("Most frequent element present in 'surface-quality' column:", df_clean['surface-quality'].value_counts().idxmax() )
print ("Most frequent element present in 'product-type' column:", df_clean['product-type'].value_counts().idxmax() )

Most frequent element present in 'steel' column: A
Most frequent element present in 'condition' column: S
Most frequent element present in 'surface-quality' column: E
Most frequent element present in 'product-type' column: C


In [20]:
# substituindo os NaN pelos elemestos mais frequentes de cada coluna
df_clean['steel'].replace(np.nan, "A", inplace=True)
df_clean['condition'].replace(np.nan, "S", inplace=True)
df_clean['surface-quality'].replace(np.nan, "E", inplace=True) 
df_clean['product-type'].replace(np.nan, "C", inplace=True) 

C:\Users\tiago\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\core\series.py:4509: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().replace(


In [21]:
# calculando a média das váriaveis
avg_formab1 = df_clean['formability'].mean(axis=0)
avg_formab2 = df_clean['strength'].mean(axis=0)
avg_formab3 = df_clean['carbon'].mean(axis=0)
avg_formab4 = df_clean['bore'].mean(axis=0)
print("Average of formability:", avg_formab1 )
print("Average of strength:", avg_formab2 )
print("Average of carbon:", avg_formab3 )
print("Average of bore:", avg_formab4 )

Average of formability: 2.2147001934235977
Average of strength: 30.82920792079208
Average of carbon: 3.492574257425743
Average of bore: 28.465346534653467


In [22]:
# substituindo os NaN pelos valores da média
df_clean['formability'].replace(np.nan, avg_formab1, inplace=True)
df_clean['strength'].replace(np.nan, avg_formab2, inplace=True)
df_clean['carbon'].replace(np.nan, avg_formab3, inplace=True)
df_clean['bore'].replace(np.nan, avg_formab4, inplace=True)

In [23]:
print(" \nTotal de itens NaN presentes nas colunas do DataFrame 'Dados' : \n\n", df_clean.isnull().sum())

 
Total de itens NaN presentes nas colunas do DataFrame 'Dados' : 

 exp_id             0
ano                0
mes                0
product-type       0
steel              0
carbon             0
hardness           0
condition          0
formability        0
strength           0
surface-quality    0
shape              0
thick              0
width              0
len                0
bore               0
recozimento        0
dtype: int64


Não existem mais colunas células vazias (NaN) neste dataframe

In [24]:
df_clean

exp_id     ano  mes product-type steel  carbon  hardness condition  \
0        2.0  2020.0  8.0            C     A       8         0         S   
1     1368.0  2020.0  8.0            C     R       0         0         S   
2     1130.0  2020.0  8.0            C     A       0        60         S   
3      402.0  2020.0  8.0            C     A       0        60         S   
4      407.0  2020.0  8.0            C     A       0        45         S   
...      ...     ...  ...          ...   ...     ...       ...       ...   
1611   145.0  2019.0  8.0            C     R       0         0         S   
1612   937.0  2019.0  8.0            C     A       0         0         S   
1613  2228.0  2019.0  8.0            C     R       0         0         S   
1614  1263.0  2019.0  8.0            C     A       0        85         S   
1615  1236.0  2019.0  8.0            C     A       0        85         S   

      formability  strength surface-quality  shape  thick   width     len  \
0          2.2147         0               G   COIL  0.700   610.0     0.0   
1          2.0000         0               E  SHEET  0.700  1300.0   762.0   
2          2.2147         0               G   COIL  2.801   385.1     0.0   
3          2.2147         0               G  SHEET  0.801   255.0   269.0   
4          2.2147         0               D   COIL  1.600   610.0     0.0   
...           ...       ...             ...    ...    ...     ...     ...   
1611       2.0000         0               E  SHEET  1.600  1500.0  4170.0   
1612       3.0000         0               E  SHEET  1.200   609.9     1.0   
1613       3.0000         0               E  SHEET  1.601   830.0   880.0   
1614       2.2147         0               G   COIL  0.400    20.0     0.0   
1615       2.2147         0               G   COIL  4.000   610.0     0.0   

      bore recozimento  
0        0        ruim  
1        0        ruim  
2        0       ideal  
3        0       ideal  
4        0     mediano  
...    ...         ...  
1611     0       ideal  
1612     0       ideal  
1613     0       ideal  
1614     0       ideal  
1615   500       ideal  

[1616 rows x 17 columns]

### 4. Salvando e exportando arquivo trabalhado

In [25]:
# por fim, salva-se o trabalho em um arquivo '.csv', que foi armazenado no repositório do Github
df_clean.to_csv('data_experimentos_CLEAN.csv')